File: cpp/0094_binary_tree_inorder_traversal.cpp
// 94. Binary Tree Inorder Traversal (Easy)
// https://leetcode.com/problems/binary-tree-inorder-traversal/
//
// Time: O(n), Space: O(h) where h = tree height

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* l, TreeNode* r) : val(x), left(l), right(r) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> stk;
        TreeNode* curr = root;

        while (curr || !stk.empty()) {
            // Go left as far as possible
            while (curr) {
                stk.push(curr);
                curr = curr->left;
            }

            // Visit node
            curr = stk.top();
            stk.pop();
            result.push_back(curr->val);

            // Move to right subtree
            curr = curr->right;
        }

        return result;
    }
};

// Helper to print vector
void print(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
}

int main() {
    Solution sol;

    // Test 1: [1,null,2,3] → [1,3,2]
    //     1
    //      \
    //       2
    //      /
    //     3
    TreeNode* t1 = new TreeNode(1, nullptr, new TreeNode(2, new TreeNode(3), nullptr));
    cout << "Test 1: ";
    print(sol.inorderTraversal(t1));  // [1, 3, 2]

    // Test 2: empty
    cout << "Test 2: ";
    print(sol.inorderTraversal(nullptr));  // []

    // Test 3: single node
    cout << "Test 3: ";
    print(sol.inorderTraversal(new TreeNode(1)));  // [1]

    // Test 4: balanced
    //     2
    //    / \
    //   1   3
    TreeNode* t4 = new TreeNode(2, new TreeNode(1), new TreeNode(3));
    cout << "Test 4: ";
    print(sol.inorderTraversal(t4));  // [1, 2, 3]

    cout << "All tests passed!" << endl;
    return 0;
}

File: rust/0094_binary_tree_inorder_traversal.rs
// 94. Binary Tree Inorder Traversal (Easy)
// https://leetcode.com/problems/binary-tree-inorder-traversal/
//
// Time: O(n), Space: O(h) where h = tree height

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution;

impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = Vec::new();
        let mut stack = Vec::new();
        let mut current = root;

        while current.is_some() || !stack.is_empty() {
            // Traverse left subtree, pushing nodes onto stack
            while let Some(node) = current {
                stack.push(Rc::clone(&node));
                current = node.borrow().left.clone();
            }

            // Visit node (pop from stack)
            if let Some(node) = stack.pop() {
                result.push(node.borrow().val);
                // Move to right subtree
                current = node.borrow().right.clone();
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn node(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }

    #[test]
    fn test_example() {
        // [1,null,2,3] → [1,3,2]
        let tree = node(1, None, node(2, node(3, None, None), None));
        assert_eq!(Solution::inorder_traversal(tree), vec![1, 3, 2]);
    }

    #[test]
    fn test_empty() {
        assert_eq!(Solution::inorder_traversal(None), vec![]);
    }

    #[test]
    fn test_single() {
        assert_eq!(Solution::inorder_traversal(node(1, None, None)), vec![1]);
    }

    #[test]
    fn test_balanced() {
        //     2
        //    / \
        //   1   3
        let tree = node(2, node(1, None, None), node(3, None, None));
        assert_eq!(Solution::inorder_traversal(tree), vec![1, 2, 3]);
    }
}

